lib/puppet/environments.rb

```diff
@@ -447,5 +451,27 @@ def expired?
         Time.now > @ttl
       end
     end
+
+    class FileMtimeEntry < Entry
+      def initialize(env, path)
+        super value
+        @mtime   = Time.now
+        @evicted = false
+        @thread  = Thread.new do
```

Comment:
potentially as many threads as we have d/ environments, but all of them are 99.999% iowait so this is no issue. computing `@evicted` in `evicted?` would mean we will do a `stat` each time we want to access environment, which is what i wanted to avoid. granted it would still be much better than current "avict always" option, but i think this implementation is pretty simple.